Problems with overloads / mypy bug

I'm having some errors with typing, if it's not a `returns` problem feel free to close this issue

**Mypy** errors:
```sh
get_stream_details.py:78: error: Argument 1 to "apply" of "Result" has incompatible type "Result[overloaded function, Any]"; expected "Result[Callable[[Stream], StreamDetails], FailureDetails]"
get_stream_details.py:78: error: Argument 1 has incompatible type "Result[StreamDetails, FailureDetails]"; expected "Maybe[Any]"
get_stream_details.py:78: error: Incompatible return value type (got "Maybe[Any]", expected "Result[StreamDetails, FailureDetails]")
get_stream_details.py:79: error: Argument 1 to "_flow" has incompatible type "Result[overloaded function, Any]"; expected "Result[overloaded function, Any]"
get_stream_details.py:79: error: Argument 1 to "from_value" of "Result" has incompatible type overloaded function; expected overloaded function
get_stream_details.py:81: error: Cannot infer type argument 1 of "_bind"
get_stream_details.py:81: error: Argument 1 to "_bind" has incompatible type "Callable[[Any], Result[Any, FailureDetails]]"; expected "Callable[[Any], Maybe[Any]]"
get_stream_details.py:81: error: Incompatible return value type (got "Result[Any, FailureDetails]", expected "Maybe[object]")
get_stream_details.py:81: error: Incompatible return value type (got "Result[Any, FailureDetails]", expected "Maybe[Any]")
get_stream_details.py:81: error: Cannot infer type argument 1 of "Success"
get_stream_details.py:82: error: Cannot infer type argument 1 of "_bind"
get_stream_details.py:82: error: Argument 1 to "_bind" has incompatible type "Callable[[Any], Result[Any, FailureDetails]]"; expected "Callable[[Any], Maybe[Any]]"
get_stream_details.py:82: error: Incompatible return value type (got "Result[Any, FailureDetails]", expected "Maybe[object]")
get_stream_details.py:82: error: Incompatible return value type (got "Result[Any, FailureDetails]", expected "Maybe[Any]")
get_stream_details.py:82: error: Cannot infer type argument 1 of "Success"

```

All the errors above are from my *return* statement in *GetStreamDetailsUsecase.\_\_call__*!

**Code**:
```python
@dataclass_json
@dataclass
class StreamDetails:
    stream: Stream
    projects: List[Project]
    ksql_stream_detailed: KSQLStreamDetailed

    @classmethod
    @curry
    def build(
        cls,
        stream: Stream,
        projects: List[Project],
        ksql_stream_detailed: KSQLStreamDetailed,
    ) -> "StreamDetails":
        return cls(stream, projects, ksql_stream_detailed)


class FindStreamByStreamId(ABC):
    @abstractmethod
    def __call__(self, stream_id: UUID) -> Result[Maybe[Stream], FailureDetails]:
        pass


class FindProjectsByStream(ABC):
    @abstractmethod
    def __call__(self, stream: Stream) -> Result[List[Project], FailureDetails]:
        pass


# "GetStreamByNameUsecase" definition
# class GetStreamByNameUsecase:
#     def __call__(
#             self, stream_name: str
#         ) -> Result[KSQLStreamDetailed, BusinessFailureDetails]:
#         pass


class GetStreamDetailsUsecase:
    def __init__(
        self,
        find_stream_by_stream_id: FindStreamByStreamId,
        find_projects_by_stream: FindProjectsByStream,
        get_stream_by_name: GetStreamByNameUsecase,
    ):
        self.__find_stream_by_stream_id = find_stream_by_stream_id
        self.__find_projects_by_stream = find_projects_by_stream
        self.__get_stream_by_name = get_stream_by_name

    def __call__(self, stream_id: UUID) -> Result[StreamDetails, FailureDetails]:
        stream = self.__find_stream_by_stream_id(stream_id).bind(
            self.__verify_if_stream_exist
        )
        partial_projects = partial(stream.bind, self.__find_projects_by_stream)
        partial_ksql = partial(
            stream.bind, lambda stream_: self.__get_stream_by_name(stream_.name)
        )

        return flow(
            Result.from_value(StreamDetails.build),
            stream.apply,
            bind(lambda to_apply: partial_projects().apply(Success(to_apply))),
            bind(lambda to_apply: partial_ksql().apply(Success(to_apply))),
        )

    def __verify_if_stream_exist(
        self, stream: Maybe[Stream]
    ) -> Result[Stream, BusinessFailureDetails]:
        return maybe_to_result(stream).alt(
            lambda __: BusinessFailureDetails(
                reason="NOT_FOUND", failure_message="Stream not found"
            )
        )
```
---
But if I change the flow to handmade flow just one error appeared:

**Mypy** error:
```sh
error: Argument 1 to "from_value" of "Result" has incompatible type overloaded function; expected overloaded function
```

**Code**:
```python
class GetStreamDetailsUsecase:
    # CODE

    def __call__(self, stream_id: UUID) -> Result[StreamDetails, FailureDetails]:
        stream = self.__find_stream_by_stream_id(stream_id).bind(
            self.__verify_if_stream_exist
        )
        partial_projects = partial(stream.bind, self.__find_projects_by_stream)
        partial_ksql = partial(
            stream.bind, lambda stream_: self.__get_stream_by_name(stream_.name)
        )

        details = Result.from_value(StreamDetails.build)
        details_added_stream = stream.apply(details)
        details_added_projects = details_added_stream.bind(lambda to_apply: partial_projects().apply(Success(to_apply)))
        details_added_ksql = details_added_projects.bind(lambda to_apply: partial_ksql().apply(Success(to_apply)))
        return details_added_ksql

    # CODE
```

Ok, looks like we found a bug in `mypy`. Here's the minimal example that works:

```python
from returns.curry import curry
from returns.result import Result

@curry
def first(a: int, b: int) -> int:
    ...

reveal_type(
    Result.from_value(first),
)
# Revealed type is 'returns.result.Result[Overload(def (a: builtins.int) -> def (b: builtins.int) -> builtins.int, def (a: builtins.int, b: builtins.int) -> builtins.int), Any]'
```

But, this does not:

```python
from returns.curry import curry
from returns.result import Result

@curry
def first(a: int, b: int, c: int) -> int:
    ...

reveal_type(
    Result.from_value(first),
)
```

Output:

```
» mypy ex.py
ex.py:9: note: Revealed type is 'returns.result.Result[Overload(def (a: builtins.int) -> Overload(def (b: builtins.int, c: builtins.int) -> builtins.int, def (b: builtins.int) -> def (c: builtins.int) -> builtins.int), def (a: builtins.int, b: builtins.int) -> def (c: builtins.int) -> builtins.int, def (a: builtins.int, b: builtins.int, c: builtins.int) -> builtins.int), Any]'
ex.py:9: error: Argument 1 to "from_value" of "Result" has incompatible type overloaded function; expected overloaded function
Found 1 error in 1 file (checked 1 source file)
```

Can you please create a new bug in mypy's issue tracker?

--- a/docs/requirements.txt
+++ b/docs/requirements.txt
@@ -13,6 +13,6 @@ pygments==2.7.1
 
 # Dependencies of our project:
 typing-extensions==3.7.4.3
-mypy==0.782
+mypy==0.790
 pytest==6.1.1
 hypothesis==5.37.1

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -60,7 +60,7 @@ typing-extensions = "^3.7"
 
 
 [tool.poetry.dev-dependencies]
-mypy = "^0.782"
+mypy = "^0.790"
 wemake-python-styleguide = "^0.14"
 flake8-pytest-style = "^1.2"
 flake8-pyi = "^20.5"

--- a/returns/contrib/hypothesis/laws.py
+++ b/returns/contrib/hypothesis/laws.py
@@ -172,7 +172,7 @@ def type_vars() -> Iterator[None]:
        for example, ``nan`` does not work for us
 
     """
-    used = types._global_type_lookup[TypeVar]  # type: ignore
+    used = types._global_type_lookup[TypeVar]
 
     def factory(thing):
         type_strategies = [
@@ -183,12 +183,12 @@ def type_vars() -> Iterator[None]:
             lambda inner: inner == inner,  # noqa: WPS312
         )
 
-    st.register_type_strategy(TypeVar, factory)  # type: ignore
+    st.register_type_strategy(TypeVar, factory)
 
     yield
 
-    types._global_type_lookup.pop(TypeVar)  # type: ignore
-    st.register_type_strategy(TypeVar, used)  # type: ignore
+    types._global_type_lookup.pop(TypeVar)
+    st.register_type_strategy(TypeVar, used)
 
 
 def _run_law(

--- a/returns/contrib/mypy/_features/flow.py
+++ b/returns/contrib/mypy/_features/flow.py
@@ -40,10 +40,10 @@ def analyze(ctx: FunctionContext) -> MypyType:
     # We use custom argument type inference here,
     # because for some reason, `mypy` does not do it correctly.
     # It inferes `covariant` types incorrectly.
-    real_arg_types = [
+    real_arg_types = tuple(
         ctx.api.expr_checker.accept(arg)   # type: ignore
         for arg in ctx.args[1]
-    ]
+    )
 
     return PipelineInference(
         ctx.arg_types[0][0],

--- a/returns/contrib/mypy/_features/kind.py
+++ b/returns/contrib/mypy/_features/kind.py
@@ -1,5 +1,5 @@
 from enum import Enum, unique
-from typing import List, Optional, Sequence
+from typing import Optional, Sequence, Tuple
 
 from mypy.checkmember import analyze_member_access
 from mypy.plugin import (
@@ -159,7 +159,7 @@ class _KindErrors(str, Enum):  # noqa: WPS600
 def _crop_kind_args(
     kind: Instance,
     limit: Optional[Sequence[MypyType]] = None,
-) -> List[MypyType]:
+) -> Tuple[MypyType, ...]:
     """Returns the correct amount of type arguments for a kind."""
     if limit is None:
         limit = kind.args[0].args  # type: ignore

--- a/returns/contrib/mypy/_typeops/inference.py
+++ b/returns/contrib/mypy/_typeops/inference.py
@@ -1,4 +1,4 @@
-from typing import List, Mapping, Optional, cast
+from typing import List, Mapping, Optional, Tuple, cast
 
 from mypy.argmap import map_actuals_to_formals
 from mypy.constraints import infer_constraints_for_callable
@@ -108,7 +108,7 @@ class PipelineInference(object):
 
     def from_callable_sequence(
         self,
-        pipeline_types: List[MypyType],
+        pipeline_types: Tuple[MypyType, ...],
         pipeline_kinds: List[int],
         ctx: CallableContext,
     ) -> MypyType:

--- a/typesafety/test_curry/test_curry/test_curry_arguments.yml
+++ b/typesafety/test_curry/test_curry/test_curry_arguments.yml
@@ -17,6 +17,48 @@
     reveal_type(multiple)  # N: Revealed type is 'Overload(def (builtins.int) -> Overload(def (builtins.int, builtins.int, d: builtins.int) -> builtins.str, def (builtins.int, builtins.int) -> def (d: builtins.int) -> builtins.str, def (builtins.int) -> Overload(def (builtins.int, d: builtins.int) -> builtins.str, def (builtins.int) -> def (d: builtins.int) -> builtins.str)), def (builtins.int, builtins.int) -> Overload(def (builtins.int, d: builtins.int) -> builtins.str, def (builtins.int) -> def (d: builtins.int) -> builtins.str), def (builtins.int, builtins.int, builtins.int) -> def (d: builtins.int) -> builtins.str, def (builtins.int, builtins.int, builtins.int, d: builtins.int) -> builtins.str)'
 
 
+- case: curry_nested_overload1
+  disable_cache: false
+  main: |
+    from typing import Generic, TypeVar
+    from returns.curry import curry
+
+    ValueType = TypeVar('ValueType')
+
+    class MyClass(Generic[ValueType]):
+        inner_value: ValueType
+
+        def __init__(self, inner_value: ValueType) -> None:
+            ...
+
+    @curry
+    def test(a: int, b: int) -> float:
+        ...
+
+    reveal_type(MyClass(test))  # N: Revealed type is 'main.MyClass[Overload(def (a: builtins.int) -> def (b: builtins.int) -> builtins.float, def (a: builtins.int, b: builtins.int) -> builtins.float)]'
+
+
+- case: curry_nested_overload2
+  disable_cache: false
+  main: |
+    from typing import Generic, TypeVar
+    from returns.curry import curry
+
+    ValueType = TypeVar('ValueType')
+
+    class MyClass(Generic[ValueType]):
+        inner_value: ValueType
+
+        def __init__(self, inner_value: ValueType) -> None:
+            ...
+
+    @curry
+    def test(a: int, b: int, c: str) -> int:
+        ...
+
+    reveal_type(MyClass(test))  # N: Revealed type is 'main.MyClass[Overload(def (a: builtins.int) -> Overload(def (b: builtins.int, c: builtins.str) -> builtins.int, def (b: builtins.int) -> def (c: builtins.str) -> builtins.int), def (a: builtins.int, b: builtins.int) -> def (c: builtins.str) -> builtins.int, def (a: builtins.int, b: builtins.int, c: builtins.str) -> builtins.int)]'
+
+
 # TODO: remove skip after this bug in `mypy` is fixed:
 # https://github.com/python/mypy/issues/8801
 - case: curry_init_magic_method

